pypga/modules/migen/spimachine.py

```diff
@@ -79,6 +79,7 @@ def __init__(self, width):
                 sr.eq(self.pdi),
                 self.sdo.eq(Mux(self.lsb_first, self.pdi[0], self.pdi[-1]))
             ),
+            #
             If(self.load,
                 sr.eq(self.pdo),
                 self.sdo.eq(Mux(self.lsb_first, self.pdo[0], self.pdo[-1]))
```

Comment:
But how on earth des this work? Again, I understand that `sdo` is set to the first or last bit of `pdo`, depending on whether `lsb_first` is true. But this line would give me the very first or the very last bit all the time. However, I would expect `sdo` to iterate over `pdo` from back to front or from front to back. And apparently it does. But why does it work?

Reply:
I  read the code and it understand it now, but I wouldn't say its well-written, because the names of the registers do not indicate what they do. As far as I understand `self.load` is only asserted once at the beginning of a transaction, and it sets `sr=pdo` (each is `width` bits wide). It also sets `sdo` to the first value it needs to assume. Then `self.shift` is asserted at every SPI cycle, setting `sr=pdi` instead. And as you see above, `pdi` is essentially a version of `sr` shifted by one bit. This way `pdi` has all the MOSI bits at the beginning, and all the MISO bits at the end. I think if this isn't clear enough we could rewrite this using different names (or for example, don't use `pdi` at all). But the important thing is that when `shift` is asserted, the data in `sr` is shifted by one bit.